File: Data Structures & Algorithms/trapping-rain-water/submission-9.py
class Solution:
    def trap(self, height: List[int]) -> int:
        left = [0] * len(height)
        right = [0] * len(height)

        max_value = 0

        for index in range(len(height)):
            max_value = max(height[index], max_value)
            left[index] = max_value
            
        
        max_value = 0

        for index in range(len(height) -1, -1, -1):
            max_value = max(height[index], max_value)
            right[index] = max_value
        
        sum = 0
        for index in range(len(left)):
            sum += min(left[index], right[index]) - height[index]
        
        return sum

File: Data Structures & Algorithms/max-water-container/submission-9.java
class Solution {
    public int maxArea(int[] heights) {
        int left = 0;
        int right = heights.length -1;
        int result = Integer.MIN_VALUE;

        while(left < right){
            int minValue = Math.min(heights[left], heights[right]);
            int distance = Math.abs(left - right);

            int currentResult = minValue * distance;
            result = Math.max(currentResult, result);

            if(heights[left] < heights[right]){
                left++;
            }else{
                right--;
            }
        }

        return result;
    }
}
